### PR TITLE
MINOR: Streams API JavaDoc improvements

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
@dguy @enothereska @guozhangwang

Some reformatting and simplifications. Also fixing some inconsistencies and errors.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -373,13 +395,13 @@ void writeAsText(final String filePath,
...
 
     /**
-     * Creates an array of {@link KStream} from this stream by branching the elements in the original stream based on
+     * Creates an array of {@link KStream} from this stream by branching the records in the original stream based on
      * the supplied predicates.
-     * Each element is evaluated against the supplied predicates, and predicates are evaluated in order.
+     * Each record is evaluated against the supplied predicates, and predicates are evaluated in order.
      * Each stream in the result array corresponds position-wise (index) to the predicate in the supplied predicates.
-     * The branching happens on first-match: An element in the original stream is assigned to the corresponding result
+     * The branching happens on first-match: An record in the original stream is assigned to the corresponding result
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: A record.
Same everywhere else An record is mentioned

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 2]
```diff
@@ -765,36 +768,59 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
-     * {@link org.apache.kafka.streams.StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is an internal
-     * generated name, and "-repartition" is a fixed suffix.
+     * {@link StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is
+     * an internal generated name, and "-repartition" is a fixed suffix.
      * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
      * <p>
-     * All data of this stream will be redistributed through the repartitioning topic by writing all records to
-     * and rereading all records from it, such that the resulting {@link KGroupedStream} is partitioned on the new key.
+     * For this case, all data of this stream will be redistributed through the repartitioning topic by writing all
+     * records to and rereading all records from it, such that the resulting {@link KGroupedStream} is partitioned
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: records to it, and reading all records from it, such that

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 3]
```diff
@@ -765,36 +768,59 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
+     * {@link #map(KeyValueMapper)}, {@link #flatMap(KeyValueMapper)}, or
+     * {@link #transform(TransformerSupplier, String...)}), and no data redistribution happened afterwards (e.g., via
+     * {@link #through(String)}) an internal repartitioning topic will be created in Kafka.
+     * This topic will be named "${applicationId}-XXX-repartition", where "applicationId" is user-specified in
+     * {@link StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is
+     * an internal generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * For this case, all data of this stream will be redistributed through the repartitioning topic by writing all
+     * records to and rereading all records from it, such that the resulting {@link KGroupedStream} is partitioned
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 4]
```diff
@@ -824,66 +849,41 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
-     * {@link #map(KeyValueMapper)}, {@link #flatMap(KeyValueMapper)}, or
-     * {@link #transform(TransformerSupplier, String...)}), and no data redistribution happened afterwards (e.g., via
-     * {@link #through(String)}) an internal repartitioning topic will be created in Kafka.
-     * This topic will be named "${applicationId}-XXX-repartition", where "applicationId" is user-specified in
-     * {@link org.apache.kafka.streams.StreamsConfig StreamsConfig} via parameter
-     * {@link org.apache.kafka.streams.StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is an internal
-     * generated name, and "-repartition" is a fixed suffix.
-     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * All data of this stream will be redistributed through the repartitioning topic by writing all records to
+     * and rereading all records from it, such that the resulting {@link KGroupedStream} is partitioned on the new key.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 5]
```diff
@@ -824,66 +849,41 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
-                                    final Serde<V> valSerde);
+    <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> selector,
+                                       final Serde<KR> keySerde,
+                                       final Serde<V> valSerde);
 
     /**
      * Join records of this stream with another {@link KStream}'s records using windowed inner equi join with default
      * serializers and deserializers.
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: record's

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 6]
```diff
@@ -920,9 +920,9 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * If this requirement is not met, Kafka Streams will automatically repartition the data, i.e., it will create an
      * internal repartitioning topic in Kafka and write and re-read the data via this topic before the actual join.
      * The repartitioning topic will be named "${applicationId}-XXX-repartition", where "applicationId" is
-     * user-specified in {@link org.apache.kafka.streams.StreamsConfig StreamsConfig} via parameter
-     * {@link org.apache.kafka.streams.StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is an internal
-     * generated name, and "-repartition" is a fixed suffix.
+     * user-specified in {@link  StreamsConfig} via parameter
+     * {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is an internal generated name, and
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: internally?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 7]
```diff
@@ -954,7 +954,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
 
     /**
      * Join records of this stream with another {@link KStream}'s records using windowed inner equi join.
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: record's

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 8]
```diff
@@ -1037,7 +1036,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * serializers and deserializers.
      * In contrast to {@link #join(KStream, ValueJoiner, JoinWindows) inner-join}, all records from this stream will
      * produce at least one output record (cf. below).
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: Ditto

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 9]
```diff
@@ -1113,7 +1111,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * Join records of this stream with another {@link KStream}'s records using windowed left equi join.
      * In contrast to {@link #join(KStream, ValueJoiner, JoinWindows, Serde, Serde, Serde) inner-join}, all records from
      * this stream will produce at least one output record (cf. below).
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 10]
```diff
@@ -1200,7 +1198,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * In contrast to {@link #join(KStream, ValueJoiner, JoinWindows) inner-join} or
      * {@link #leftJoin(KStream, ValueJoiner, JoinWindows) left-join}, all records from both streams will produce at
      * least one output record (cf. below).
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: record's

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 11]
```diff
@@ -1277,7 +1274,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * In contrast to {@link #join(KStream, ValueJoiner, JoinWindows, Serde, Serde, Serde) inner-join} or
      * {@link #leftJoin(KStream, ValueJoiner, JoinWindows, Serde, Serde, Serde) left-join}, all records from both
      * streams will produce at least one output record (cf. below).
-     * The join is a primary key join with join attribute {@code thisKStream.key == otherKStream.key}.
+     * The join is computed on the records key with join attribute {@code thisKStream.key == otherKStream.key}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: record's

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 12]
```diff
@@ -1615,99 +1617,103 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
-     * {@link GlobalKTable} state.
-     * In contrast, processing {@link GlobalKTable} input records will only update the internal {@link GlobalKTable} state and
-     * will not produce any result records.
+     * This is done by performing a lookup for matching records in the <em>current</em> internal {@link GlobalKTable}
+     * state.
+     * In contrast, processing {@link GlobalKTable} input records will only update the internal {@link GlobalKTable}
+     * state and will not produce any result records.
      * <p>
-     * For each {@link KStream} record whether or not it finds an corresponding record in {@link GlobalKTable} the provided
+     * For each {@link KStream} record that finds an corresponding record in {@link GlobalKTable} the provided
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: an -> a

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 13]
```diff
@@ -1615,99 +1617,103 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
-     * <p>
-     * For each {@link KStream} record that finds an corresponding record in {@link GlobalKTable} the provided
-     * {@link ValueJoiner} will be called to compute a value (with arbitrary type) for the result record.
-     * The key of the result record is the same as the key of this {@link KStream}
+     * This is done by performing a lookup for matching records in the <em>current</em> internal {@link GlobalKTable}
+     * state.
+     * In contrast, processing {@link GlobalKTable} input records will only update the internal {@link GlobalKTable}
+     * state and will not produce any result records.
+     * <p>
+     * For each {@link KStream} record whether or not it finds an corresponding record in {@link GlobalKTable} the
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: an -> a

**Review by dguy — CHANGES_REQUESTED** [review]
Some spelling/grammar changes

**Comment by mjsax** [comment]
Updated.

**Comment by guozhangwang** [comment]
Merged to trunk and 0.10.2.

**Pull request closed without merging**